packages/runtime/runtime-utils/src/summarizerNodeContract.ts

```diff
@@ -0,0 +1,78 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { ISnapshotTree } from "@fluidframework/protocol-definitions";
+import {
+    IGraphNode,
+    ISummarizeInternalResult,
+    ISummarizerNode,
+    ISummarizerNodeConfig,
+    ISummarizerNodeWithGC,
+} from "@fluidframework/runtime-definitions";
+import { SummarizerNode } from "./summarizerNode";
+import { ReadAndParseBlob } from "./summarizerNodeUtils";
+import { SummarizerNodeWithGC } from "./summarizerNodeWithGc";
+
+export { ReadAndParseBlob } from "./summarizerNodeUtils";
+
+export interface ISummarizerNodeRootContract {
```

Comment:
SummarizerNode should implement this interface. If not, this should not be exported.

Reply:
This design was intentional: to minimize surface area of runtime-definitions while achieving same level of type safety, and prevent circular dependencies within these files.

Consider the following constraints:
- I don't want to export SummarizerNode itself, because it is an implementation which then requires all interfaces/types that are _protected_ and public to be exported.
- I need 2 separate interfaces: 1 - `ISummarizerNode` for all nodes in the tree, and 2 - `IRootSummarizerNode` for the root node only.  Previously I didn't need `IRootSummarizerNode`, because only root functions were exposed on SummarizerNode -it wasn't meant to be extended.
- Think of each file as a module, I want to avoid circular module dependencies within this package.
    - I must export `IRootSummarizerNode` (to use it in ContainerRuntime)
    - I must not export `SummarizerNode` (to hide internals)
    - I must export a function which will create an `IRootSummarizerNode` (to use it in ContainerRuntime)
    - **Option 1:** Add a 5th file containing only definitions for `IRootSummarizerNode` which is also exported
    - **Option 2:** Move `IRootSummarizerNode` definition into utils file, and then import->export the definitions in contract file
    - **Option 3:** Move `IRootSummarizerNode` to runtime-definitions (this is reasonable)
    - **Option 4:** Leverage TypeScript structural typing to prevent the dependency from (summarizerNodeContract.ts -> summarizerNode.ts)

Option (3) seems good, but I wanted to keep it minimal as possible, and I knew that I had to export the `createRootSummarizerNode` functions anyway, which meant adding an equivalent contract file, since those should not go in the definitions package.

Without touching runtime-definitions, I would choose option (4) over (1) or (2) to avoid excessive files and directly passthrough import->exporting the definitions. In my opinion all 3 options have pros/cons, but none are wrong.

The drawback for not implementing it, is that we get an error in the summarizerNodeContract.ts file instead of in the summarizerNode.ts file when we make a change to SummarizerNode that violates it. It's pretty minor, but I'm okay to make the change to option (1) or (2) if you prefer?

Reply:
Great write-up, and really nice thought process.  Do we all agree Option 3 is best, and are there plans to make that move soon?  I can't tell if

> I wanted to keep it minimal as possible

refers to this PR or the architecture as a whole.

Reply:
And since we're letting this be loosely coupled for now, I'd suggest/request a nice comment on this interface and/or the `SummarizerNode` class calling out the connection since it's not explicit to the compiler.

Reply:
I was referring to the architecture. I imagine runtime-definitions has more depending on it than runtime-utils, so I wanted to minimize the size of runtime-definitions. That's my hesitation, especially if we never end up needing these interfaces anywhere else.

Of (1), (2), (4), I could go either way, this was just my initial choice: 4<5 files, 2<4 passthrough import->exports, but tsc error in summarizerNodeContract.ts instead of summarizerNode.ts

Reply:
Changed to option (2) in latest PR. Let me know what you think.
After doing so, the contract file became only passthrough import->exports which is kind of nice I guess.

Reply:
Not in the scope of this PR but I was thinking about the architecture as a whole.
Should ISummarizerNode, ISummarizerNodeWithGC and other related exports move here from runtime-definitions so that all summarizer node related functionality is in one place? It looks like those are generic enough to stay in runtime-definitions but just curious as to what you think.

Reply:
Oh and I like the current iteration + Tony's suggestion of moving summarize node related stuff to a folder.

Reply:
Some of the definitions must live in runtime-definitions, because they bleed into the contract between runtime/container <-> runtime/dataStore (along the boundary of DataStoreContext to DataStoreRuntime).  Those should _not_ live in runtime-utils, since defintions should not depend on utils.

Reply:
Makes sense, thanks!